File: examples/web/src/nav/nav.ts
import type { FeedbackKind, GpRoute, NavState } from "./types";

export function homeNav(): NavState {
  return { stack: ["home"], scroll: {}, promptOpen: false };
}

export function isRoute(value: unknown): value is GpRoute {
  return value === "home" || value === "settings" || value === "about" || value === "feedback";
}

export function normalizeStack(raw: readonly unknown[]): GpRoute[] {
  const out: GpRoute[] = [];
  for (const item of raw) {
    if (!isRoute(item)) continue;
    if (out.length === 0 && item !== "home") out.push("home");
    if (out[out.length - 1] === item) continue;
    out.push(item);
  }
  if (out.length === 0) return ["home"];
  if (out[0] !== "home") return ["home", ...out];
  return out;
}

export function current(state: NavState): GpRoute {
  const stack = normalizeStack(state.stack);
  return stack[stack.length - 1] ?? "home";
}

export function isHome(state: NavState): boolean {
  return current(state) === "home" && normalizeStack(state.stack).length <= 1;
}

export function canPop(state: NavState): boolean {
  return !isHome(state);
}

export function setPrompt(state: NavState, open: boolean): NavState {
  if (state.promptOpen === open) return state;
  return { ...state, promptOpen: open };
}

export function push(state: NavState, route: GpRoute, kind?: FeedbackKind): NavState {
  if (route === "home") {
    if (isHome(state) && state.feedbackKind === undefined) return state;
    return { ...state, stack: ["home"], feedbackKind: undefined };
  }
  if (current(state) === route) return state;
  return {
    ...state,
    stack: [...normalizeStack(state.stack), route],
    feedbackKind: route === "feedback" ? (kind ?? "bug") : undefined,
  };
}

export function pop(state: NavState): NavState {
  if (state.promptOpen) return { ...state, promptOpen: false };
  if (!canPop(state)) return state;
  const stack = normalizeStack(state.stack).slice(0, -1);
  const next: GpRoute[] = stack.length === 0 ? ["home"] : stack;
  const top = next[next.length - 1] ?? "home";
  return {
    ...state,
    stack: next,
    feedbackKind: top === "feedback" ? state.feedbackKind : undefined,
  };
}

File: examples/android/app/src/main/java/dev/foss/goldenpath/ui/nav/Nav.kt
package dev.foss.goldenpath.ui.nav

object Nav {
    fun home(): NavState = NavState(stack = listOf(GpRoute.Home))

    fun normalizeStack(raw: List<GpRoute>): List<GpRoute> {
        val out = mutableListOf<GpRoute>()
        for (item in raw) {
            if (out.isEmpty() && item != GpRoute.Home) out.add(GpRoute.Home)
            if (out.lastOrNull() == item) continue
            out.add(item)
        }
        if (out.isEmpty()) return listOf(GpRoute.Home)
        if (out.first() != GpRoute.Home) return listOf(GpRoute.Home) + out
        return out
    }

    fun current(state: NavState): GpRoute = normalizeStack(state.stack).lastOrNull() ?: GpRoute.Home

    fun isHome(state: NavState): Boolean =
        current(state) == GpRoute.Home && normalizeStack(state.stack).size <= 1

    fun canPop(state: NavState): Boolean = !isHome(state)

    fun setPrompt(state: NavState, open: Boolean): NavState =
        if (state.promptOpen == open) state else state.copy(promptOpen = open)

    fun push(state: NavState, route: GpRoute, kind: FeedbackKind? = null): NavState {
        if (route == GpRoute.Home) {
            if (isHome(state) && state.feedbackKind == null) return state
            return state.copy(stack = listOf(GpRoute.Home), feedbackKind = null)
        }
        if (current(state) == route) return state
        return state.copy(
            stack = normalizeStack(state.stack) + route,
            feedbackKind = if (route == GpRoute.Feedback) kind ?: FeedbackKind.Bug else null,
        )
    }

    fun pop(state: NavState): NavState {
        if (state.promptOpen) return state.copy(promptOpen = false)
        if (!canPop(state)) return state
        val next = normalizeStack(state.stack).dropLast(1).ifEmpty { listOf(GpRoute.Home) }
        val top = next.last()
        return state.copy(
            stack = next,
            feedbackKind = if (top == GpRoute.Feedback) state.feedbackKind else null,
        )
    }

    fun recordScroll(state: NavState, route: GpRoute, y: Int): NavState {
        if (route == GpRoute.Home) return state
        val n = y.coerceAtLeast(0)
        return state.copy(scroll = state.scroll + (route to n))
    }

    fun restoreScroll(state: NavState, route: GpRoute): Int =
        state.scroll[route]?.coerceAtLeast(0) ?: 0
}
